File: egg-niuke-training/26-valid-parentheses.py
"""
NC52 有效括号序列
https://www.nowcoder.com/practice/37548e94a270412c8b9fb85643c8ccc2

给出一个仅包含字符'(',')','{','}','['和']',的字符串，判断给出的字符串是否是合法的括号序列。

栈的应用

时间复杂度：O(n)
空间复杂度：O(n)
"""

def is_valid_parentheses(s):
    """
    有效括号序列
    """
    if not s:
        return True
    
    stack = []
    mapping = {')': '(', '}': '{', ']': '['}
    
    for char in s:
        if char in mapping:
            # 右括号
            if not stack or stack.pop() != mapping[char]:
                return False
        else:
            # 左括号
            stack.append(char)
    
    return len(stack) == 0

# 测试
if __name__ == "__main__":
    test_cases = [
        "()",
        "()[]{}",
        "(]",
        "([)]",
        "{[]}",
        ""
    ]
    
    for s in test_cases:
        result = is_valid_parentheses(s)
        print(f"'{s}' 是否有效: {result}")

File: egg-niuke-training/26-valid-parentheses.cpp
/**
 * NC52 有效括号序列
 * https://www.nowcoder.com/practice/37548e94a270412c8b9fb85643c8ccc2
 * 
 * 给出一个仅包含字符'(',')','{','}','['和']',的字符串，判断是否是合法的括号序列。
 * 
 * 时间复杂度：O(n)
 * 空间复杂度：O(n)
 */

#include <iostream>
#include <string>
#include <stack>
#include <unordered_map>
using namespace std;

class Solution {
public:
    bool isValid(string s) {
        if (s.empty()) return true;
        
        stack<char> st;
        unordered_map<char, char> mapping = {
            {')', '('},
            {'}', '{'},
            {']', '['}
        };
        
        for (char c : s) {
            if (mapping.count(c)) {
                // 右括号
                if (st.empty() || st.top() != mapping[c]) {
                    return false;
                }
                st.pop();
            } else {
                // 左括号
                st.push(c);
            }
        }
        
        return st.empty();
    }
};

int main() {
    Solution solution;
    
    vector<string> testCases = {
        "()",
        "()[]{}",
        "(]",
        "([)]",
        "{[]}",
        ""
    };
    
    for (const string& s : testCases) {
        bool result = solution.isValid(s);
        cout << "'" << s << "' 是否有效: " << (result ? "是" : "否") << endl;
    }
    
    return 0;
}

